src/service/url-expander/expander.js

```diff
@@ -0,0 +1,264 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {rethrowAsync} from '../../log';
+
+/** Rudamentary parser to handle nested Url replacement. */
+export class Expander {
+
+  /**
+   * Link this instance of parser to the calling UrlReplacment
+   * @param {VariableSource} variableSource the keywords to replace
+   */
+  constructor(variableSource) {
+    this.variableSource_ = variableSource;
+  }
+
+  /**
+   * take the template url and return a promise of its evaluated value
+   * @param {string} url url to be substituted
+   * @param {!Object<string, *>=} opt_bindings additional one-off bindings
+   * @param {!Object<string, boolean>=} opt_whiteList Optional white list of names
+   *     that can be substituted.
+   * @return {!Promise<string>}
+   */
+  expand(url, opt_bindings, opt_whiteList) {
+    if (!url.length) {
+      return Promise.resolve(url);
+    }
+    const expr = this.variableSource_
+        .getExpr(opt_bindings, /*opt_ignoreArgs */ true);
+
+    const matches = this.findMatches_(url, expr);
+    // if no keywords move on
+    if (!matches.length) {
+      return Promise.resolve(url);
+    }
+    const mergedPositions = this.eliminateOverlaps_(matches, url,
+        opt_whiteList);
+    return this.parseUrlRecursively_(url, mergedPositions, opt_bindings);
+  }
+
+  /**
+   * Structures the regex matching into the desired format
+   * @param {string} url url to be substituted
+   * @param {RegExp} expression regex containing all keywords
+   * @return {!Object<string, string|number>} array of objects representing
+   *  matching keywords
+   */
+  findMatches_(url, expression) {
+    const matches = [];
+    url.replace(expression, (match, name, startPosition) => {
+      const length = match.length;
+      const stopPosition = length + startPosition - 1;
+      const info = {
+        start: startPosition,
+        stop: stopPosition,
+        name,
+        length,
+      };
+      matches.push(info);
+    });
+    return matches;
+  }
+
+  /**
+   * Takes array of all matching variable keyword locations and returns only those that we
+   * want to evaluate. In the case of overlapping keywords it choose the longer. It also
+   * excludes protected keywords.
+   * @param {!Array<Object<string, string|number>>} matches array of objects representing
+   *  matching keywords
+   * @param {string} url The url to be expanded.
+   * @return {!Object<string, string|number>}
+   */
+  eliminateOverlaps_(matches, url, opt_whiteList) {
+    if (!matches.length) { return null; }
+    // longest keywords have priority over shorter
+    // if same length alphabetical precidence
+    matches.sort((a, b) => {
```

Comment:
might worth to extract the sorting function to constants.

Reply:
irrelevant since we confirmed regex behavior